### PR TITLE
Escape '<' in JSON strings

### PR DESCRIPTION
**Pull request by jamesdonoh**
Resolves https://github.com/bbc/simorgh-infrastructure/issues/923 (and other misc issues)

### Overall change
Ensures that the less-than symbol in JSON strings is escaped (using a Unicode escape sequence) before being output inside script tags where it could be incorrectly picked up by the HTML parser.

Full credit to https://uploadcare.com/blog/vulnerability-in-html-design/ for exploring this problem in detail and proposing a practical solution.

### Code changes
- adds a utility function that serialises a JS object 'safely' for output in a script tag
- use this function where ever we currently perform this task

### Impact
The problem affects two places I can see in our codebase

#### `SIMORGH_DATA`
Because we output the data that will be used to rehydrate the page on the client using `dangerouslySetInnerHTML`
Screenshot 2020-03-29 at 11 47 47

#### LinkedData component
Because `react-helmet` uses `dangerouslySetInnerHTML` internally.
Screenshot 2020-03-29 at 11 47 36

### Existing mitigations/related work
- Includes fetching/rendering: https://github.com/bbc/simorgh/pull/5894
- Social embeds: https://github.com/bbc/simorgh/pull/5983
- Converting escaped characters in paragraph blocks back to the literal form: https://github.com/bbc/simorgh/blob/latest/src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/paragraph/helpers.js#L8-L24
- Escape less than/greater than/ampersands in the text of internal links: https://github.com/bbc/simorgh/blob/latest/src/app/routes/cpsAsset/getInitialData/convertToOptimoBlocks/blocks/internalLinks/index.js#L53-L57

### Further reading
- HTML spec on permitted content inside script tags: https://html.spec.whatwg.org/multipage/scripting.html#restrictions-for-contents-of-script-elements

### Alternative solutions
We could use the escaping method described in the HTML spec above, this is slightly more verbose and it is not really clear whether it is preferable to the solution here

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added labels to this PR for the relevant pod(s) affected by these changes
- [x] I have assigned this PR to the Simorgh project

**Testing:**
- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`) 
- [no] This PR requires manual testing

### CODE REVIEW

File: src/app/lib/utilities/serialiseForScript/index.test.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      expect(serialiseForScript({ text: 'me < you' })).toBe(
+        '{"text":"me \\u003c you"}',
+      );
+    });
+
+    it('an object with a less-than symbol in a property name', () => {
+      expect(serialiseForScript({ '<blink>': 'hello!' })).toBe(
+        '{"\\u003cblink>":"hello!"}',
+      );
+    });
```
Comment (andrewscfc) [inline src/app/lib/utilities/serialiseForScript/index.test.js:34]: Should we add a test for a JSON object that contains a script, just to be clear what this intends to do with a _'real'_ example:
```javascript
expect(serialiseForScript({ "html": "<div>An include</div><script>console.log('An include');</script>" }))
  .toBe('{ "html": "\\u003cdiv>An include\\u003c/div>\\u003cscript>console.log('An include');\\u003c/script>"
}');
```

File: src/app/lib/utilities/serialiseForScript/index.test.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  });
+
+  describe('unexpected input', () => {
+    it('null', () => {
+      expect(serialiseForScript(undefined)).toBeUndefined();
```
Comment (simonsinclair) [inline src/app/lib/utilities/serialiseForScript/index.test.js]: Should this be?:
```js
expect(serialiseForScript(null)).toBeUndefined();
```

Comment (simonsinclair) [inline src/app/lib/utilities/serialiseForScript/index.test.js]: Is this meant to be?:
```js
expect(serialiseForScript(null)).toBeNull();
```

**Comment by andrewscfc** [comment]
cc @rhenshaw56 

Just quickly merged this branch into ~our includes branch~ latest (now with includes merged) and this fix seems to work for includes too: https://github.com/bbc/simorgh/compare/latest...includes-safer-serialisation. I removed our adhoc encoding solution for includes.

We still have one outstanding issue with a server mismatch error but that was the case with our adhoc encoding solution we used previously:
image

I think we are still getting this error because we do a server render of the include then, rehydrate with the escaped page data which will be a little bit different due to the escaped `<`. I would also expect to see this for social embeds too @simonsinclair ?

Reintroducing `suppressHydrationWarning` here would hide the warning: https://github.com/bbc/simorgh/blob/237d6820156d265052749bc44313b742586bd7b4/src/app/containers/Include/index.jsx#L23

Obviously, we still need to go through the includes implementation and adjust the tests etc to remove the adhoc encoding tests, this can be done in another issue.

**Review by andrewscfc — APPROVED** [review]
Happy to take this to test. I'd suggest:
- Run our full cypress suite against it
- Allow teams to complete any manual checks for their specific pages

File: src/app/lib/utilities/serialiseForScript/index.test.js (hunk 2 above)
Comment (jamesdonoh) [reply src/app/lib/utilities/serialiseForScript/index.test.js]: Oops yes, copy-paste error. Ta.

Comment (jamesdonoh) [reply src/app/lib/utilities/serialiseForScript/index.test.js]: Actually, the result is `undefined` due to the short-circuit OR. I don't think it makes much difference in practice though.

**Review by rhenshaw56 — APPROVED** [review]

**Comment by simonsinclair** [comment]
> We still have one outstanding issue with a server mismatch error but that was the case with our adhoc encoding solution we used previously:
> image
> 
> I think we are still getting this error because we do a server render of the include then, rehydrate with the escaped page data which will be a little bit different due to the escaped `<`. I would also expect to see this for social embeds too @simonsinclair ?

No warning here, @andrewscfc.

**Review by simonsinclair — APPROVED** [review]

File: src/app/lib/utilities/serialiseForScript/index.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+// output in a `script` tag without confusing the HTML parser
+//
+// The approach used is simply to replace literal less-than characters ('<') with
+// an equivalent Unicode escape sequence. For more detail see this blog post:
+// https://uploadcare.com/blog/vulnerability-in-html-design/
+//
+// See also:
+// https://html.spec.whatwg.org/multipage/scripting.html#restrictions-for-contents-of-script-elements
+
+export default obj => obj && JSON.stringify(obj).replace(/</g, '\\u003c');
```
Comment (EinsteinNjoroge) [inline src/app/lib/utilities/serialiseForScript/index.js]:
```suggestion
export default (obj) => obj && JSON.stringify(obj).replace(/</g, '\\u003c');
```
A fix for the eslint error that causing jenkins to fail

**Review by EinsteinNjoroge — COMMENTED** [review]
Jenkins is failing due to eslint

File: src/app/lib/utilities/serialiseForScript/index.js (hunk 3 above)
Comment (jamesdonoh) [reply src/app/lib/utilities/serialiseForScript/index.js]: Thanks @EinsteinNjoroge!

**Comment by j-pendlebury** [comment]
Jenkins still failing, this time due to the bundle size:
`Bundle size for Russian is too large at 794 kB`

**Comment by jamesdonoh** [comment]
😩

**Pull request merged**